File: cli/src/commands/validate.ts
import { Command } from "commander";
import * as path from "path";
import chalk from "chalk";
import { findACSRoot } from "../../../reference-impl/typescript/parser";
import { validateProject } from "../../../reference-impl/typescript/validator";

export function validateCommand(): Command {
  const cmd = new Command("validate");
  cmd
    .description("Validate all ACS files in the current project")
    .option("-p, --path <dir>", "Path to project root (default: cwd)")
    .action((opts: { path?: string }) => {
      const start = opts.path ? path.resolve(opts.path) : process.cwd();
      const root = findACSRoot(start);

      if (!root) {
        console.error(
          chalk.red("✗ No .agents/main.yaml found in this directory or any parent.")
        );
        process.exit(1);
      }

      console.log(chalk.dim(`Validating: ${root}`));
      const result = validateProject(root);

      if (result.errors.length === 0) {
        console.log(chalk.green("✓ ACS project is valid."));
        process.exit(0);
      } else {
        console.log(chalk.red(`✗ Found ${result.errors.length} error(s):\n`));
        for (const err of result.errors) {
          console.log(`  ${chalk.red("•")} ${err}`);
        }
        process.exit(1);
      }
    });

  return cmd;
}

File: packages/python/acs_cli/commands/validate.py
"""acs validate — validate all ACS files in the current project."""

import os
import sys
from pathlib import Path

_HERE = os.path.dirname(__file__)
_REPO_ROOT = os.path.abspath(os.path.join(_HERE, "..", "..", "..", ".."))
_REFIMPL = os.path.join(_REPO_ROOT, "reference-impl", "python")
if _REFIMPL not in sys.path:
    sys.path.insert(0, _REFIMPL)

from acs_validator import validate_project  # noqa: E402
from acs_parser import find_acs_root  # noqa: E402


def run(args) -> int:
    start = Path(args.path) if args.path else Path.cwd()
    root = find_acs_root(start)

    if root is None:
        print("\033[31m✗ No .agents/main.yaml found in this directory or any parent.\033[0m", file=sys.stderr)
        return 1

    print(f"\033[2mValidating: {root}\033[0m")
    result = validate_project(root)
    errors = result.get("errors", [])

    if not errors:
        print("\033[32m✓ ACS project is valid.\033[0m")
        return 0
    else:
        print(f"\033[31m✗ Found {len(errors)} error(s):\033[0m\n")
        for err in errors:
            print(f"  \033[31m•\033[0m {err}")
        return 1
